Storage, Container and Product structures

N-Wouda opened the pull request:
1. Storage ('Warehouse', 'Factory', or 'Store'). A place where goods are stored for long-term.
2. Container ('Rack', 'Barrel', 'Box'). A storage box within a Storage facility. Contains products.
3. Product. Has properties (perishability? Price? Size?), and is contained within Containers.

Feel free to comment and discuss. Once finished, this closes #3.

N-Wouda committed:
Updated gitignore with default mac/windows settings

N-Wouda committed:
Skeletons for the container, product and storage classes

N-Wouda committed:
Initial implementation for the Product class

N-Wouda commented on classes/product.js:
```diff
@@ -0,0 +1,41 @@
+     * @param {int} quantity - The amount of this product
+     * @param {int} perishable - The number of days a perishable good remains fresh. 0 for infinite.
+     */
+    constructor(name, quantity, perishable = 0)
```
We should do something with product dimensions (else the Container cannot know its maximum contents). Could be weight, size, volume? Any ideas?

lukko replied:
Maybe have like a generic measurement in 'units'.
For example: a container has a capacity of 100 units. A bike equals 5 units and a chair equals 2 units, then one container can contain:
- 20 bikes or
- 16 bikes and 10 chairs

Has the advantage that we don't have to deal with length, width and height of products for now.

N-Wouda replied:
Seems reasonable to me

N-Wouda committed:
Container skeleton, and Luc's suggestion for unitless size

N-Wouda commented on classes/product.js:
```diff
@@ -0,0 +1,52 @@
+    {
+        this.name = String(name);
+        this.quantity = parseInt(quantity);
+        this.size = parseFloat(size);
```
@lukko Does this seem reasonable? I think this is pretty much what you suggested :).

N-Wouda commented on classes/container.js:
```diff
@@ -0,0 +1,48 @@
+class Container
```
TODO: A way to remove products from a container.

Have this be done automatically when the quantity on any product drops to zero (could be done when calculating the total used capacity).

N-Wouda committed:
Type checking, inheritance and Warehouse (Storage) skeleton

N-Wouda committed:
Forgot toString on Warehouse

I do need to think about the semantic problem of having a 'container'
extend Storage (quite general), but also 'Warehouse' (quite specific.
Maybe delegate the 'name' property downwards, or have specific classes
for each type (like Box, Rack, Warehouse, Factory)?

N-Wouda commented on classes/storage.js:
```diff
@@ -0,0 +1,42 @@
+    {
+        var availableCapacity = this.capacity - this._usedCapacity();
+
+        if (!availableCapacity) {
```
Should check for `<= 0`, not just zero.

N-Wouda committed:
Finished initial implementations for Storage, Container, and Product

Using two 'core' classes, ProductCore and StorageCore, we can maintain
some level of hierarchy, while retaining flexibility.

Bugfix on precendence in boolean evaluation (instanceof)

No idea how that got in there :)

Corrected a JSDoc

N-Wouda commented:
@JanVeen @VinnieBoySlim I do not have write access to the repo, but I think this is part is finished for now. Feel free to review and if suitable, merge the PR.

Also, could I perhaps be given write access so I can merge PRs myself? :)

VincentVelthuizen commented:
@N-Wouda Since I think it is appropriate for you to have elevated rights I have just assigned you admin privilege on this repo. The default for the team is still read, we could make it write, what do you think?

N-Wouda commented:
> The default for the team is still read, we could make it write, what do you think?

'Write' allows merging pull requests, which may not be desirable. Not that I think anyone would intentionally push faulty code, but accidents happen. Read should suffice for opening PRs. Then merging can be done by anyone with Write access.

N-Wouda commented:
If no further comment/review I am merging this later today.

lukko commented:
Why don't you use camelcase for file names?

N-Wouda commented:
> Why don't you use camelcase for file names?

I generally follow something along the lines of [this SO answer](http://stackoverflow.com/questions/7273316/what-is-the-javascript-filename-naming-convention#answer-7273448) for JS code, but more generally, [this also holds true](http://stackoverflow.com/questions/13511227/url-structure-lower-case-vs-upper-case):

> The path part is or is not case sensitive, depending on the server environment and server. Typically Windows machines are case insensitive, while Linux machines are case sensitive. This means that you should stick to lowercase or you risk introducing a bug that's really hard to hunt down (mismatched case that doesn't matter on the dev server).
